packages/skygear-core/src/imageprocessing.ts

```diff
@@ -0,0 +1,162 @@
+import { encodeQueryComponent } from "./url";
+
+/**
+ * @public
+ */
+export interface ImageProcessingPipelineBuilderResizeOptions {
+  scalingMode?: "lfit" | "mfit" | "fill" | "pad" | "fixed";
+  targetWidth?: number;
+  targetHeight?: number;
+  longerSide?: number;
+  shorterSide?: number;
+  color?: string;
+}
+
+type Operation = FormatOperation | QualityOperation | ResizeOperation;
+
+interface FormatOperation {
+  op: "format";
+  format: "jpg" | "png" | "webp";
+}
+
+interface QualityOperation {
+  op: "quality";
+  absoluteQuality: number;
+}
+
+interface ResizeOperation extends ImageProcessingPipelineBuilderResizeOptions {
+  op: "resize";
+}
+
+/**
+ * @public
+ */
+export class ImageProcessingPipelineBuilder {
+  private _ops: Operation[];
+
+  constructor() {
+    this._ops = [];
+  }
+
+  format(format: "jpg" | "png" | "webp"): this {
+    this._ops.push({
+      op: "format",
+      format,
+    });
+    return this;
+  }
+
+  quality(absoluteQuality: number): this {
+    this._ops.push({
+      op: "quality",
+      absoluteQuality,
+    });
+    return this;
+  }
+
+  resize(options: ImageProcessingPipelineBuilderResizeOptions): this {
+    this._ops.push({
+      op: "resize",
+      ...options,
+    });
+    return this;
+  }
+
+  getName(): string {
+    return "pipeline";
+  }
+
+  getValue(): string {
+    const parts = ["image"];
+    for (const op of this._ops) {
+      switch (op.op) {
+        case "format":
+          parts.push(renderFormat(op));
+          break;
+        case "quality":
+          parts.push(renderQuality(op));
+          break;
+        case "resize":
+          parts.push(renderResize(op));
+          break;
+        default:
+          break;
+      }
+    }
+    return parts.join("/");
+  }
+
+  setToURLString(urlStr: string): string {
+    const fragmentIndex = urlStr.indexOf("#");
```

Comment:
Can we switch to using URL API? It seems it has more or less same level of support as fetch.

Reply:
This is a trouble of supporting three platforms.

URL API is part of the Web API. Node.js has it but we have to import it from the `url` module. React Native does not have it at all.

I tried to find good URL and URLSearchParams polyfill but had no luck. Some of them only polyfill URL or URLSearchParams. If both are polyfilled, they polyfill the global which cannot be used by us because we are writing a library.

If we require the user to provide such polyfill in React Native, it is not a good user experience.

So I finally decided to implement `setToURLString` without using any library or API.

Reply:
Does [`whatwg-url`](https://github.com/jsdom/whatwg-url) fits the requirement?

Reply:
It has many dependencies

Reply:
You're right. Let's keep it as is for now. For the record, I just tested in NodeJS and it does have a global `URL` class in version 10, so the problem would be on React Native.